### PR TITLE
+str #17399 add boilerplate remover for fan-in junctions

### PR DESCRIPTION
**Pull request by agolubev**
Please verify if implementation approach is good enough. I'm still not sure about return type for sink mirrored version.
If it's Ok I'll need to add following: 
- docs
- java API
- sink tests
- more source tests
- check formatting for boilerplate code

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Refs #17399

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala [hunk 1]
```diff
@@ -153,6 +153,29 @@ class SourceSpec extends AkkaSpec {
       gotten.toSet should ===(Set(0, 1, 2, 3, 4))
       out.expectComplete()
     }
+
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala]: No tests for divide?

Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala]: Had intention to add them (as well as docs and java api) after positive feedback on implementation approach.
Still if you need this to see API usage I add them right away.

Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala]: I completely understand that rationale, but if you look at it from the readers' point of view, there's really nothing that verifies that it does work and if so what it does.

**Comment by drewhk** [comment]
I think this approach looks good. You can also change now the implementation of Source.concat to use this feature instead of creating the graph manually. So please proceed with tests :)

**Comment by agolubev** [comment]
Yeah, thanks, will proceed.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/SinkApply.scala.template [hunk 2]
```diff
@@ -47,4 +47,21 @@ trait SinkApply {
...
+   * Combine several sinks with fun-out strategy like `Broadcast` or `Balance` and returns `Sink`.
+   */
+  def combine[T, U]([#output1: Sink[U, Any]#])(strategy: Int => Graph[UniformFanOutShape[T, U], Unit]): Sink[T, Unit] =
+    Sink([#output1#])(([#M1#]) => ()) { implicit b ⇒
+      ([#s1#]) ⇒
+        import FlowGraph.Implicits._
+        val d = b.add(strategy(1))
+
+        [#d.out(0) ~> s1.inlet#
+        ]
```
Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SinkApply.scala.template]: why newline here?

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template [hunk 3]
```diff
@@ -49,4 +49,22 @@ trait SourceApply {
   
   ]
 
+  [2..#/**
+   * Combines several sources with fun-in strategy like `Merge` or `Concat` and returns`Source`.
```
Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template]: missing space in "returns`Source`"

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/SinkApply.scala.template (hunk 2 above)
Comment (agolubev) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SinkApply.scala.template]: Idea was to visually divide boilerplate section with out ~> in.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template [hunk 4]
```diff
@@ -49,4 +49,22 @@ trait SourceApply {
...
+      ([#s1#]) ⇒
+        import FlowGraph.Implicits._
+        val c = b.add(strategy(1))
+
+        [#s1 ~> c.in(0)#
+        ]
+
+        SourceShape(c.out)
+    })
+  #
```
Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template]: I wonder if using templating for this is the optimal solution (given the code duplication and scaling),
wdyt about something like the following?

~~~scala

 def combine[T, U](first: Source[T, _], second: Source[T, _], rest: Source[T, _]*)(strategy: Int ⇒ Graph[UniformFanInShape[T, U], Unit]): Source[U, Unit] =
    Source.wrap(FlowGraph.partial() { implicit b ⇒
      import FlowGraph.Implicits._

      val c = b.add(strategy(rest.size + 2))

      @tailrec def addRest(idx: Int, i: Iterator[Source[T, _]]): SourceShape[U] =
        if (i.hasNext) {
          i.next() ~> c.in(idx)
          addRest(idx + 1, i)
        } else SourceShape(c.out)

      first ~> c.in(0)
      second ~> c.in(1)
      addRest(2, rest.iterator)
    })

~~~

Comment (agolubev) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/SourceApply.scala.template]: Make sense. Will connect all in-outs with help of internal function

**Comment by agolubev** [comment]
Ok, I moved Scala API to be non boilerplate as was suggested.

For Java API it would be something like this:
```scala
@varargs def combine[T, U](strategy: function.Function[java.lang.Integer, Graph[UniformFanOutShape[T, U], Unit]],
                           outputs: Sink[U, _]*): Sink[T, Unit] =
  new Sink[T, Unit](scaladsl.Sink.combine(outputs.map(_.asScala): _*)(num ⇒ strategy.apply(num)))
```
However we have two obstacles here:
* varargs must be as the last argument. It makes us rearrange function attributes and create difference between Java and Scala API (still it's not critical).
* also there is an issue in Scala compiler https://issues.scala-lang.org/browse/SI-8743 which is fixed only in 2.11.3 - it hardly can be workarounded with the given API.

Consequently I left Java API in boilerplate as it was.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala [hunk 5]
```diff
@@ -252,4 +253,22 @@ object Sink extends SinkApply {
   def actorSubscriber[T](props: Props): Sink[T, ActorRef] =
     new Sink(new ActorSubscriberSink(props, DefaultAttributes.actorSubscriberSink, shape("ActorSubscriberSink")))
 
+  /**
+   * Combine several sinks with fun-out strategy like `Broadcast` or `Balance` and returns `Sink`.
+   */
+  def combine[T, U](sinks: Sink[U, _]*)(strategy: Int ⇒ Graph[UniformFanOutShape[T, U], Unit]): Sink[T, Unit] =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala]: To enforce that it is not used with less than 2, this should (IMO) be:

~~~scala
def combine[T, U](sink1: Sink[U, _], sink2:Sink[U, _], sinkN: Sink[U, _]*)(strategy: Int ⇒ Graph[UniformFanOutShape[T, U], Unit]): Sink[T, Unit] = …
~~~

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala [hunk 6]
```diff
@@ -89,6 +91,25 @@ class SinkSpec extends AkkaSpec {
       }
     }
 
+    "combine from many inputs with simplified API" in {
+      val probes = Seq.fill(3)(TestSubscriber.manualProbe[Int]())
+      val sink = Sink.combine(Sink(probes(0)), Sink(probes(1)), Sink(probes(2)))(Broadcast(_))
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala]: There's no test for when an empty varargs is used. or a single Sink.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala (hunk 5 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala]: Yeah, make sense.

**Comment by agolubev** [comment]
Ok, it seems like build verifies that Java DSL is equal to Scala DSL. Should I align them or add exception to test?

**Comment by viktorklang** [comment]
Aligning the Java API sounds like the best option IMO.

**Comment by drewhk** [comment]
Due to a Scala issue, we couldn't make the varargs version proper for Java, and therefore we removed it (you can still use a Collection).

**Comment by drewhk** [comment]
But Java API equivalents for combine is necessary, but you can add an exception on the varargs version, creating a Collection based alternative in Java.

**Comment by agolubev** [comment]
Yeah, just committed with List in Java API, as it seems it's the closest to scala Seq. (http://www.scala-lang.org/docu/files/collections-api/collections_46.html)

**Comment by ktoso** [comment]
It's moving in the right direction, thanks @agolubev! Though we won't include it in the 1.0 today, we'll get back to it and get it into 1.1

**Comment by drewhk** [comment]
Hi @agolubev can you rebase this PR? i would like to merge it as it seems ready to me.

**Comment by agolubev** [comment]
Ok, will do

**Comment by agolubev** [comment]
Seems like HttpEntitySpec failed. 
Let me know if something relates to my commit

**Comment by agolubev** [comment]
@drewhk it's 5 days passed - perhaps it will require rebase once again.

**Comment by drewhk** [comment]
PLS BUILD

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala (hunk 6 above)
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala]: +1, that test case is important

File: akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala [hunk 7]
```diff
@@ -107,6 +108,26 @@ object Sink extends SinkApply {
     Flow[T].map(f).toMat(Sink.ignore)(Keep.right).named("foreachSink")
 
   /**
+   * Combine several sinks with fun-out strategy like `Broadcast` or `Balance` and returns `Sink`.
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala]: "fan-out"

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala [hunk 8]
```diff
@@ -226,6 +226,15 @@ object Source {
       case s: Source[T, M] ⇒ s
       case other           ⇒ new Source(scaladsl.Source.wrap(other))
     }
+
+  /**
+   * Combines several sources with fun-in strategy like `Merge` or `Concat` and returns `Source`.
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: "fan-in"

**Comment by drewhk** [comment]
Looking good, please add a small section to the doc pages, then I merge this.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala (hunk 6 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala]: now combine definition is like
```def combine[T, U](first: Sink[U, _], second: Sink[U, _], rest: Sink[U, _]*)``` 
so I'll add test for two sinks

**Comment by agolubev** [comment]
Ok, added small doc and rebased

**Comment by agolubev** [comment]
TcpSpec "shutdown if Flows have not been subscribed" case failed

**Comment by agolubev** [comment]
@drewhk, I finished with this

File: akka-docs-dev/rst/scala/stream-graphs.rst [hunk 9]
```diff
@@ -152,6 +152,20 @@ For defining a ``Flow[T]`` we need to expose both an inlet and an outlet:
 
 .. includecode:: code/docs/stream/StreamPartialFlowGraphDocSpec.scala#flow-from-partial-flow-graph
 
+Combining Sources and Sinks with simplified API
+-----------------------------------------------
+
+There is simplified API you can use to combine sources and sinks with junctions like: ``Broadcast[T]``, ``Balance[T]``,
+``Merge[In]`` and ``Concat[A]``. With method ``combine`` you don't need to connect junction
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/stream-graphs.rst]: ".. without the need for using the Graph DSL. The combine method takes care of constructing the necessary graph underneath."

File: akka-docs-dev/rst/scala/stream-graphs.rst [hunk 10]
```diff
@@ -152,6 +152,20 @@ For defining a ``Flow[T]`` we need to expose both an inlet and an outlet:
...
+with sources or sinks. It will construct simple graph for you. In following example we combine two sources into
+one (fun-in):
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/stream-graphs.rst]: fan-in

**Comment by drewhk** [comment]
Java documentation section is missing yet.

**Comment by agolubev** [comment]
build failed with following:
```
maven-out] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:testCompile (default-testCompile) on project akka-docs-java-lambda: Compilation failure
[maven-out] [ERROR] required: Graph<UniformFanInShape<Integer,Integer>,BoxedUnit>
[maven-out] [ERROR] found:    Graph
[maven-out] [ERROR] /mnt/akka/jenkinsakka/localhome-a0/workspace/pr-validator-per-commit-jenkins/akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/StreamPartialFlowGraphDocTest.java:[168,25] error: cannot find symbol
```
this is strange because when I put problematic test case into SinkTest and set target java to 1.8 it works ok.

**Comment by agolubev** [comment]
@drewhk any ideas?

**Comment by agolubev** [comment]
Sorry for so many build failures. It took me sometime to figure out how to work properly with lambda samples and docs for java

**Comment by agolubev** [comment]
@drewhk, it's ready

**Comment by drewhk** [comment]
Almost ;) Please squash the two remaining commits, and then I will merge after validation. Awesome work @agolubev !

**Comment by agolubev** [comment]
build failed with TcpSpec issue. Fix is in PR #18369

**Comment by drewhk** [comment]
PLS BUILD

**Pull request merged**